### PR TITLE
feat: autonomous tool routing from plain chat (#74)

### LINKED ISSUE #74: Autonomous Tool Routing from Plain Chat
Feature:
Allow Jarvis to call tools autonomously from plain chat (non-slash input), by classifying intent and selecting the right tool plan.

Scope:
- Add intent classifier for plain text input (chat/research/pm/delegate/self-review)
- If non-chat intent is detected with high confidence, route to corresponding agent with tool access
- Keep safe fallback to CHAT mode when confidence is low
- Log selected route + confidence in response footer for transparency

DoD:
- Plain prompt like "???????? ?????? ??? redrobot" routes to actionable workflow instead of CHAT-only
- Low-confidence prompts remain in safe chat mode
- Unit/integration tests for routing edge cases
- Documentation updated with behavior and examples

### PR DESCRIPTION
**Pull request by Osasuwu**
## Summary

- **Intent routing**: plain text messages are now classified and routed to the right skill automatically. Keyword matching runs first (free, instant); Haiku LLM fallback handles ambiguous inputs. Every response includes a transparency footer (`[jarvis] route: /triage | confidence: 0.83 | mode: intent-routed`).
- **Dispatcher architecture**: `dispatcher.py` rewritten with unified `dispatch_skill()` entry point supporting handler-based and prompt-based skills. `telegram.py`, `main.py`, `registry.py` simplified accordingly. Background-task detection in Telegram now pre-routes intent, so `"улучши себя"` correctly runs `/self-improve` in a background thread.
- **Skill wiring** (#75, #76): `handle()` entry points added to `self_review.py`, `self_improve.py`, `delegate.py`; SKILL.md frontmatter updated with `handler`, `model`, `max_budget_usd`, `background` fields.
- **Tests**: 39 unit tests in `tests/test_intent_router.py` covering keyword routing, async LLM fallback (mocked), confidence thresholds, edge cases, and transparency formatting.

## Test plan

- [x] `python -m compileall src/` — no errors
- [x] `pytest tests/ -v` — 39/39 passed
- [x] Smoke test: imports, `discover_skills()`, `_should_run_in_background("улучши себя") == True`
- [ ] Manual: send plain text like "triage my issues" to Telegram bot, verify it routes to `/triage` with footer
- [ ] Manual: send greeting like "привет", verify it stays in chat mode

## Risk / rollback

Low risk. Intent routing is purely additive — slash commands bypass it entirely. If routing misfires, confidence threshold (0.78 keyword / 0.70 LLM) keeps false positives low. To disable routing, remove the 10-line block in `dispatch_skill()`.

Closes #74

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/jarvis/dispatcher.py [hunk 1]
```diff
@@ -17,62 +35,318 @@ class UnsupportedCommandError(ValueError):
...
+    if skill.handler:
+        try:
+            handler = _resolve_handler(skill.handler)
+        except (ImportError, AttributeError) as exc:
+            return SkillResult(text=f"Handler import failed: {exc}", success=False)
+
+        try:
+            result: SkillResult = await handler(config, args)
+        except Exception as exc:
+            return SkillResult(text=f"Handler error: {exc}", success=False)
```
[the remark this answers is not in the document]
Comment (Osasuwu) [reply src/jarvis/dispatcher.py:308]: Fixed in 4d1fd7e — dispatcher now calls record_execution() after any handler that returns non-zero cost/tokens, using the session_id passed into dispatch_skill().

File: src/jarvis/delegate.py [hunk 2]
```diff
@@ -354,3 +354,35 @@ async def delegate_issue(
...
+        return SkillResult(text=str(exc), success=False)
+
+    result = await delegate_issue(
+        repo,
+        issue_number,
+        max_budget_usd=min(0.30, config.budget.per_query_usd),
+        session_id="delegate-dispatch",
+        daily_budget_usd=config.budget.per_day_usd,
+        per_query_usd=config.budget.per_query_usd,
+    )
```
[the remark this answers is not in the document]
Comment (Osasuwu) [reply src/jarvis/delegate.py:376]: Fixed in 4d1fd7e — removed hardcoded 0.30 cap and constant session_id. Now uses config.budget.per_query_usd as the budget cap and generates a unique uuid-based session_id per call.

File: src/handlers/telegram.py [hunk 3]
```diff
@@ -289,10 +237,15 @@ def run_telegram_loop(config: RuntimeConfig) -> int:
...
             user_input = _resolve_user_input(parsed.text)
 
-            if is_delegation_command(user_input):
-                _send_message(token, parsed.chat_id, "[jarvis] Delegation request accepted. Running in background...")
+            if _should_run_in_background(user_input):
+                command = user_input.split(maxsplit=1)[0]
+                _send_message(
+                    token,
+                    parsed.chat_id,
+                    f"[jarvis] {command} started. Running in background...",
```
[the remark this answers is not in the document]
Comment (Osasuwu) [reply src/handlers/telegram.py]: Fixed in 4d1fd7e — for slash commands the start message shows the command token; for plain-text routed inputs (e.g. Ukrainian/Russian phrases) it shows the generic 'Background task started' instead of the first word of the input.

File: tests/test_intent_router.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+
+import asyncio
+from unittest.mock import AsyncMock, patch
+
+import pytest
+import sys
+import os
+
+# Ensure src/ is on the path.
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))
```
[the remark this answers is not in the document]
Comment (Osasuwu) [reply tests/test_intent_router.py]: Fixed in 4d1fd7e — removed sys.path.insert block. pytest resolves imports via pythonpath = ["src"] in pyproject.toml.

File: src/main.py [hunk 5]
```diff
@@ -5,13 +5,15 @@
...
-from jarvis.costs import check_daily_budget, record_execution
 from jarvis.config import load_config
-from jarvis.delegate import delegate_issue, parse_delegate_args
-from jarvis.dispatcher import UnsupportedCommandError, build_prompt_for_user_input
-from jarvis.executor import execute_query
+from jarvis.dispatcher import (
+    UnsupportedCommandError,
+    build_prompt_for_user_input,
+    dispatch_skill,
+    discover_skills,
```
[the remark this answers is not in the document]
Comment (Osasuwu) [reply src/main.py]: Fixed in 4d1fd7e — removed unused discover_skills import.

File: src/jarvis/dispatcher.py [hunk 6]
```diff
@@ -1,8 +1,26 @@
...
+Two execution modes:
+1. Prompt-based (default): SKILL.md body is sent to Claude as a prompt
+2. Handler-based: a Python async function handles the full pipeline
+
+Adding a new skill = create skills/<name>/SKILL.md with frontmatter. Done.
+"""
 from __future__ import annotations
 
+import importlib
+import re
```
[the remark this answers is not in the document]
Comment (Osasuwu) [reply src/jarvis/dispatcher.py]: Fixed in 4d1fd7e — removed unused import re.

File: src/jarvis/delegate.py [hunk 7]
```diff
@@ -354,3 +354,35 @@ async def delegate_issue(
         coding_summary=coding_result.summary,
         cost_usd=cost, input_tokens=in_tok, output_tokens=out_tok,
     )
+
+
+async def handle(config: "RuntimeConfig", args: str) -> "SkillResult":
+    """Skill handler entry point — called by dispatcher auto-discovery."""
+    from jarvis.config import RuntimeConfig as _RC  # noqa: WPS433, F811
```
[the remark this answers is not in the document]
Comment (Osasuwu) [reply src/jarvis/delegate.py]: Fixed in 4d1fd7e — removed the unused RuntimeConfig import inside handle().

**Comment by Osasuwu** [comment]
Closing: autonomous tool routing was built on the old dispatcher pattern which was deleted in the reboot (PR #82). Intent routing is now handled natively by Claude Code.
